Deprecate COSINE VectorSimilarity function

Pulkitg64 opened the pull request:
### Description

Tries to solve #13281 by deprecating `COSINE` VectorSimilarity function for Vector Search.

Pulkitg64 commented:
`TestInt8HnswBackwardsCompatibility.java` and `TestBasicBackwardsCompatibility` test cases failing for Lucene versions older than `LUCENE_10_0_0`. For older Lucene version, we try to read Index `directory` from resources folder. For example for Lucene-9.10.0, the directory path is located at `lucene/backward-codecs/build/resources/test/org/apache/lucene/backward_index/int8_hnsw.9.10.0.zip`.

As per my understanding the segment present at above directory already has `KNNVectorField` with MAXIMUM_INNER_PRODUCT as vector similarity function and in above test we try to change similarity function to `DOT_PRODUCT` due to which there is conflict happening when adding more docs to it and hence these test cases failing.

benwtrent commented on lucene/core/src/java/org/apache/lucene/index/VectorSimilarityFunction.java:
```diff
@@ -61,24 +60,6 @@ public float compare(byte[] v1, byte[] v2) {
-   * should only use this function if you need to preserve the original vectors and cannot normalize
-   * them in advance. The similarity score is normalised to assure it is positive.
-   */
-  COSINE {
-    @Override
-    public float compare(float[] v1, float[] v2) {
-      return Math.max((1 + cosine(v1, v2)) / 2, 0);
-    }
-
-    @Override
-    public float compare(byte[] v1, byte[] v2) {
-      return (1 + cosine(v1, v2)) / 2;
-    }
-  },
```
You cannot do this. These enumerations have default ordinals. We need to either:

 - detach these ordinal IDs from the implementation
 - Switch from an enumerator to an `id->implementation` idea. 

Even if deprecated, Lucene needs to be able to read indices from 9x -> 10. So, cosine cannot go away on read.

Pulkitg64 replied:
Thanks @benwtrent for commenting!

> These enumerations have default ordinals

Yeah, I guess that's why some tests are failing. For Lucene9.x the ordinal is getting mapped to `MAXIMUM_INNER_PRODUCT` which was earlier mapping to `COSINE`.

> We need to either detach these ordinal IDs from the implementation or Switch from an enumerator to an id-implementation idea.

I am not sure how to do this but let me try to give it a shot.

> Even if deprecated, Lucene needs to be able to read indices from 9x -> 10. So, cosine cannot go away on read.

Oh okay, that means we need to remove support for COSINE just from indexing side not from searching side?

uschindler replied:
I am a bit confused why this still happens. @ChrisHegarty already submitted a PR to have a separate mapping from file format's internal ordinal numbers to enum constants.
Maybe this was not yet applied, let me check it out.

uschindler replied:
See here: #13119

Maybe the change there was not the whole story. If we have another places where original numbers of enum constants are used, let's use same approach per codec version.

Maybe we need to only change the historical codecs?

uschindler replied:
Anyways, we have to keep the constant in 10.x for reading old indexes.

uschindler replied:
I found the issue. #13119 is fine. The problem was that inside the codec the ordinal to enum constant was changed, which is backwards incompatible.

Basically the mapping in the codec needs to be changed while writing to reject the deprecated constant.

jpountz commented:
> Oh okay, that means we need to remove support for COSINE just from indexing side not from searching side?

Correct. Practically, this means keeping the enum constant but marking it as deprecated, and modifying IndexingChain to fail using COSINE on 10.x Lucene indexes.

It would be nice to also add a new file format under lucene/codecs that does cosine similarity, like we just added for hamming distance. This could help with the migration for some users.

Pulkitg64 commented:
Thanks @jpountz for confirming and suggestions.

uschindler commented on lucene/core/src/java/org/apache/lucene/codecs/lucene94/Lucene94FieldInfosFormat.java:
```diff
@@ -302,7 +301,6 @@ private static VectorSimilarityFunction getDistFunc(IndexInput input, byte b) th
       List.of(
           VectorSimilarityFunction.EUCLIDEAN,
           VectorSimilarityFunction.DOT_PRODUCT,
-          VectorSimilarityFunction.COSINE,
```
Here is exactly the problem. There must be a null entry, so the ordinals in index format stay alive.

ChrisHegarty replied:
++ The reason for the list here is exactly to support and allow such evolution and separation from the enum ordinals. A null here trivially need to a change of List implementation type.   And tests need to be carefully updated.

uschindler replied:
of course, null does not work with default lists :-) Also downstream the code needs to check for a null value.

In any case we can later also go back to a Map<Integer,VectorSimilarity> or to an array.

Pulkitg64 replied:
Thanks @uschindler @ChrisHegarty  for the pointers.

benwtrent commented on lucene/core/src/java/org/apache/lucene/index/VectorSimilarityFunction.java:
```diff
@@ -61,24 +60,6 @@ public float compare(byte[] v1, byte[] v2) {
   },
 
-  /**
-   * Cosine similarity. NOTE: the preferred way to perform cosine similarity is to normalize all
```
I think it would be good to deprecate this first, get all the internal paths updated to not use it. 

Then the deprecation can be merged and backported. Then work on removing it from main.

Mostly because the deprecation path and internal code usage will all have to occur anyways :)

benwtrent commented on lucene/core/src/java/org/apache/lucene/util/quantization/ScalarQuantizedRandomVectorScorer.java:
```diff
@@ -30,23 +28,6 @@
 public class ScalarQuantizedRandomVectorScorer
     extends RandomVectorScorer.AbstractRandomVectorScorer<byte[]> {
 
-  public static float quantizeQuery(
```
Even in Lucene 10, we will need to support quantizing a query sent to a Lucene index that uses Cosine.

benwtrent commented on lucene/core/src/java/org/apache/lucene/index/VectorSimilarityFunction.java:
```diff
@@ -61,24 +60,6 @@ public float compare(byte[] v1, byte[] v2) {
-      return (1 + cosine(v1, v2)) / 2;
-    }
-  },
-
```
I think this will get tricky. If you remove this ordinal (I think it would be wise to deprecate first, work through the warnings, etc. get everything passing, etc.), field info & knn vector format readers will need to somehow distinguish between "Max inner product" (whose ordinal decreases to cosine's) and "cosine but in an old index version".

Maybe on write, we adjust the ordinal dynamically, knowing what they are (thus correcting MIP's ordinal on write, allowing for readers to distinguish).

benwtrent reviewed (commented):
Left a larger review. 

I think it would be prudent to first attempt to deprecate and handle internal usage, etc. After that is merged and backported, another PR can be made to fully remove it from main only on writes.

At least, that would make reviewing & testing this simpler :)

Pulkitg64 commented on lucene/backward-codecs/src/test/org/apache/lucene/backward_index/TestBasicBackwardsCompatibility.java:
```diff
@@ -98,8 +99,14 @@ public class TestBasicBackwardsCompatibility extends BackwardsCompatibilityTestB
   private static final String KNN_VECTOR_FIELD = "knn_field";
   private static final FieldType KNN_VECTOR_FIELD_TYPE =
-      KnnFloatVectorField.createFieldType(3, VectorSimilarityFunction.COSINE);
+      KnnFloatVectorField.createFieldType(3, VectorSimilarityFunction.DOT_PRODUCT);
```
Make sense.

Pulkitg64 committed:
Mark COSINE Vector Similarity Function as Deprecated

Pulkitg64 committed:
Mark COSINE Vector Similarity Function as Deprecated

Pulkitg64 committed:
tidy fixes

Pulkitg64 committed:
tidy fixes

Pulkitg64 commented:
Thanks @benwtrent for all the feedback.
 I have raised another revision just to mark the COSINE function as deprecated. I didn't find any internal usages of this function.
 In a followup PR, I will remove COSINE from writes for LUCENE version >=10.0.0

Pulkitg64 committed:
Added entry to CHANGES.txt

benwtrent commented:
@Pulkitg64 could you deprecate also `VectorUtil.cosine`? That is technically a public API.

Pulkitg64 committed:
Mark VectorUtil cosine function as deprecated

Pulkitg64 committed:
Mark VectorUtil cosine function as deprecated

benwtrent commented on lucene/CHANGES.txt:
```diff
@@ -102,6 +102,8 @@ API Changes
   Additionally, deprecated methods have been removed from ByteBuffersIndexInput, BooleanQuery and others. Please refer
   to MIGRATE.md for further details. (Sanjay Dutt)
 
+* GITHUB#13281: Mark COSINE VectorSimilarityFunction as deprecated. (Pulkit Gupta)
+
```
This should be under 9.11 as we will backport the deprecation as a way to warn users that it is going away in 10.0

Pulkitg64 replied:
Yes, Make sense. Will change in next revision
Thanks for mentioning.

Pulkitg64 replied:
Done

Pulkitg64 committed:
Move change from 10.0 to 9.11.0

Pulkitg64 committed:
Move change from 10.0 to 9.11.0

ChrisHegarty reviewed (approved):
LGTM, though you it would be good get another reviewers approval too.

Pulkitg64 commented:
Hi @benwtrent, could we merge this change, if everything looks good to you?

Pulkitg64 committed:
Merge branch 'main' into cosine-deprecate

donnerpeter committed:
hunspell: speed up "compress"; minimize the number of the generated entries; don't even consider "forbidden" entries anymore (#13429)

Hunspell: speed up "compress"; minimize the number of the generated entries; don't even consider "forbidden" entries anymore

pseudo-nymous committed:
Fixes failing test case for TestOrdinalMap.testRamBytesUsed (#13421)

stefanvodita committed:
Allow users to retrieve counts from taxo association facets (#13414)

Add a count field to LabelAndValue

benwtrent committed:
Add next minor version 9.12.0

jpountz committed:
Fix TestIndexWriterOnError.testIOError failure. (#13436)

Pull request #13406 inadvertly broke Lucene's handling of tragic exceptions by
stopping after the first `DocValuesProducer` whose `close()` calls throws an
exception, instead of keeping calling `close()` on further producers in the
list.

This moves back to the previous behavior.

Closes #13434

ChrisHegarty committed:
SimpleText[Float|Byte]VectorValues::scorer should return null when the vector values is empty (#13444)

This commit ensures that SimpleText[Float|Byte]VectorValues::scorer returns null when the vector values is empty, as per the scorer javadoc. Other KnnVectorsReader implementations have specialised empty implementations that do similar, e.g. OffHeapFloatVectorValues.EmptyOffHeapVectorValues. The VectorScorer interface in new in Lucene 9.11, see #13181

An existing test randomly hits this, but a new test has been added that exercises this code path consistently. It's also useful to verify other KnnVectorsReader implementations.

benwtrent committed:
Add new dynamic confidence interval configuration to scalar quantized format (#13445)

When int4 scalar quantization was merged, it added a new way to dynamically calculate quantiles.

However, when that was merged, I inadvertently changed the default behavior, where a null confidenceInterval would actually calculate the dynamic quantiles instead of doing the previous auto-setting to 1 - 1/(dim + 1).

This commit formalizes the dynamic quantile calculate through setting the confidenceInterval to 0, and preserves the previous behavior for null confidenceIntervals so that users upgrading will not see different quantiles than they would expect.

easyice committed:
Fix test failure on TestPoint#testEqualsAndHashCode (#13433)

jpountz committed:
Add prefetching support to stored fields. (#13424)

This adds `StoredFields#prefetch(int)`, which mostly delegates to
`IndexInput#prefetch`. Callers can take advantage of this API to parallelize
I/O across multiple stored documents by first calling `StoredFields#prefetch`
on all doc IDs before calling `StoredFields#document` on all doc IDs.

I added a cache of recently prefetched blocks to the default codec, in order to
avoid prefetching the same block multiple times in a short period of time. This
felt sensible given that doc ID reordering via recursive graph bisection or
index sorting are likely to result in search results being clustered.

msokolov committed:
mention KnnVectorsFormat in o.a.l.codecs package javadocs (#13448)

Co-authored-by: Michael Sokolov <sokolovm@amazon.com>

easyice committed:
Avoid unnecessary memory allocation in PackedLongValues#Iterator (#13439)

We always allocate a long array of page size for a new PackedLongValues#Iterator instance, which is not necessary when packing a small number of values. this is more evident in the scenario of high-frequency flush operations

ioanatia committed:
Rewrite newSlowRangeQuery to MatchNoDocsQuery when upper > lower (#13425)

jpountz committed:
Use `ReadAdvice#NORMAL` on files that have a forward-only access pattern. (#13450)

This applies to files where performing readahead could help:
 - Doc values data (`.dvd`)
 - Norms data (`.nvd`)
 - Docs and freqs in postings lists (`.doc`)
 - Points data (`.kdd`)

Other files (KNN vectors, stored fields, term vectors) keep using a `RANDOM`
advice.

jpountz committed:
Add prefetching for doc values and norms. (#13411)

This follows a similar approach as postings and only prefetches the first page
of data.

I verified that it works well for collectors such as `TopFieldCollector`, as
`IndexSearcher` first pulls a `LeafCollector`, then a `BulkScorer` and only
then starts feeding the `BulkScorer` into the `LeafCollector`. So the
background I/O for the `LeafCollector` which will prefetch the first page of
doc values and the background I/O for the `BulkScorer` will run in parallel.

benwtrent commented:
@Pulkitg64 sorry for radio silence, I was out of office and swamped with other things. If you could move the change log to 9.12, I can merge and backport.

bugmakerrrrrr committed:
Implement Weight#count for vector values in the FieldExistsQuery (#13322)

* implement Weight#count for vector values

* add change log

* apply review comment

* apply review comment

* changelog

* remove null check

ChrisHegarty committed:
Update Gradle wrapper to 8.8 - supports Java 22 (#13453)

This commit updates the Gradle wrapper to 8.8, which has support for Java 22.

dweiss committed:
Update WrapperDownloader to accept java 22 and correct deprecated new URL API #13458

dweiss committed:
Add a github workflow that checks common (and less common) gradle tasks when gradle version is changed (#13456)

dweiss committed:
Java 22 has been released, so drop -ea from smoketester gh workflow matrix.

jpountz committed:
Add test for ghost fields to BaseKnnVectorQueryTestCase. (#13455)

iamsanjay committed:
Removed Scorer#getWeight (#13440)

If Caller requires Weight then they have to keep track of Weight with which Scorer was created in the first place instead of relying on Scorer.

Closes #13410

benwtrent committed:
DOAP changes for release 9.11.0

iverase committed:
Merge related HashMaps in FieldInfos#FieldNumbers into one map (#13460)

Merges all immutable attributes in FieldInfos.FieldNumbers into one hashmap saving memory when 
writing big indices. Fixes an exotic bug when calling clear where not all attributes were cleared.

benwtrent committed:
Sync CHANGES for 9.11.0

benwtrent committed:
Add back-compat indices for 9.11.0

mayya-sharipova committed:
MultiTermQuery return null for ScoreSupplier (#13454)

MultiTermQuery return null for ScoreSupplier if there are no terms in an index that
 match query terms.

With the introduction of PR #12156 we saw degradation in performance of bool queries
where one of the mandatory clauses is a TermInSetQuery with query terms not present in
the field. Before for such cases TermsInSetQuery returned null for ScoreSupplier which
would shortcut the whole bool query.

This PR adds ability for MultiTermQuery to return null for ScoreSupplier if a field
doesn't contain any query terms.

Relates to PR #12156

iverase committed:
Move entry in CHANGES.txt

original-brownbear committed:
Reduce the heap use of BKDReader instances (#13464)

We consume a lot of memory for the `indexIn` slices. If `indexIn` is of
type `MemorySegmentIndexInput` the overhead of keeping loads of slices
around just for cloning is far higher than the extra 12b per reader this
adds (the slice description alone often costs a lot).
In a number of Elasticsearch example uses with high segment counts I
investigated, this change would save up to O(GB) of heap.

Commit:
on README.md, make links to CONTRIBUTING.md more prominent, and demote link to dev-docs

Commit:
fix fumble-finger

msokolov committed:
Document how to make tests run faster in IntelliJ (#13466)

also make links to CONTRIBUTING.md more prominent, and demote link to dev-docs

benwtrent committed:
Add int8_hnsw backcompat index creawtion to dev tools scripts (#13465)

Commit:
clarify that IntelliJ UI varies across platforms

Pulkitg64 committed:
Move changes from 9.11.0 to 9.12.0

Pulkitg64 committed:
Merge remote-tracking branch 'origin/cosine-deprecate' into cosine-deprecate

Pulkitg64 commented:
I somehow messed up with my branch while pulling changes from `main` branch. @benwtrent I have created a new PR for marking the cosine VectorSImilarity function as deprecated.

https://github.com/apache/lucene/pull/13473

Sorry for the mess!
